File: backend/internal/models/openclaw_upgrade_lab.go
package models

import "time"

type OpenClawUpgradeLabRun struct {
	ID                  int64      `db:"id,primarykey,autoincrement" json:"id"`
	ActorUserID         *int       `db:"actor_user_id" json:"actor_user_id,omitempty"`
	Status              string     `db:"status" json:"status"`
	Phase               string     `db:"phase" json:"phase"`
	BaselineImageRef    string     `db:"baseline_image_ref" json:"baseline_image_ref"`
	BaselineImageDigest string     `db:"baseline_image_digest" json:"baseline_image_digest"`
	SourceDeployment    string     `db:"source_deployment" json:"source_deployment"`
	InstanceIDsJSON     *string    `db:"instance_ids_json" json:"-"`
	TargetImageRef      *string    `db:"target_image_ref" json:"target_image_ref,omitempty"`
	RolloutID           *int64     `db:"rollout_id" json:"rollout_id,omitempty"`
	BeforeJSON          *string    `db:"before_json" json:"-"`
	ChecksJSON          *string    `db:"checks_json" json:"-"`
	ErrorCode           *string    `db:"error_code" json:"error_code,omitempty"`
	ErrorMessage        *string    `db:"error_message" json:"error_message,omitempty"`
	CreatedAt           time.Time  `db:"created_at" json:"created_at"`
	UpdatedAt           time.Time  `db:"updated_at" json:"updated_at"`
	FinishedAt          *time.Time `db:"finished_at" json:"finished_at,omitempty"`
}

func (OpenClawUpgradeLabRun) TableName() string { return "openclaw_upgrade_lab_runs" }

File: frontend/src/types/openClawUpgradeLab.ts
import type { Instance } from './instance';
import type { RuntimeUpgradeDetails } from './runtimePool';

export interface OpenClawUpgradeLabRun {
  id: number;
  actor_user_id?: number;
  status: string;
  phase: string;
  baseline_image_ref: string;
  baseline_image_digest: string;
  source_deployment: string;
  target_image_ref?: string;
  rollout_id?: number;
  error_code?: string;
  error_message?: string;
  created_at: string;
  updated_at: string;
  finished_at?: string;
}

export interface OpenClawUpgradeLabCheck {
  name: string;
  passed: boolean;
  expected?: string;
  actual?: string;
  message?: string;
}

export interface OpenClawUpgradeLabBaselineEvidence {
  instance_id: number;
  project_file_count: number;
  manual_project_file_count: number;
  session_count: number;
  user_message_count: number;
  assistant_message_count: number;
  interactive_user_message_count: number;
  project_sha256: string;
  session_catalog_sha256: string;
  session_source_sha256: string;
}

export interface OpenClawUpgradeLabView {
  run?: OpenClawUpgradeLabRun;
  instance_ids: number[];
  instances: Instance[];
  checks: OpenClawUpgradeLabCheck[];
  rollout?: RuntimeUpgradeDetails;
  baseline_tag: string;
  baseline_evidence: OpenClawUpgradeLabBaselineEvidence[];
}
